src/main.rs

```diff
@@ -775,37 +846,19 @@ fn backtrace(
                 .get(address)
                 .map(|symbol| symbol.name())
                 .unwrap_or("???");
-            println!("{:>4}: {}", frame_index, name);
+            backtrace_display_str.push_str(&format!("{:>4}: {}\n", frame_index, name));
             frame_index += 1;
         }
 
-        // on hard fault exception entry we hit the breakpoint before the subroutine prelude (`push
-        // lr`) is executed so special handling is required
-        // also note that hard fault will always be the first frame we unwind
-        if top_exception.is_none() {
-            top_exception = Some(if pc & !THUMB_BIT == vector_table.hard_fault & !THUMB_BIT {
-                // HardFaultTrampoline
-
-                let stack_overflow = if let Some(sp_ram_region) = sp_ram_region {
-                    // NOTE stack is full descending; meaning the stack pointer can be `ORIGIN(RAM) +
-                    // LENGTH(RAM)`
-                    let range = sp_ram_region.range.start..=sp_ram_region.range.end;
-                    !range.contains(&sp)
-                } else {
-                    log::warn!(
-                        "no RAM region appears to contain the stack; cannot determine if this was a stack overflow"
-                    );
-
-                    false
-                };
-
-                TopException::HardFault { stack_overflow }
-            } else {
-                TopException::Other
-            });
+        if print_backtrace {
+            // we need to print everything we've collected up until now, otherwise the
+            // debug level logs won't match up
+            print!("{}", backtrace_display_str);
```

Comment:
personally I'd put the `\n` not inside the str and use `println!("{}", bt_without_newline)`, but *again* a matter of taste

Reply:
in principle I totally agree; the reason I went for for print! here is that `print_backtrace` is somposed differently depending on which branches are taken beforehand so figuring out which line will be the last to be added would make things more complicated in the current state of the code I think